packages/statemanager/src/stateManager.ts

```diff
@@ -157,43 +318,41 @@ export class DefaultStateManager extends BaseStateManager implements StateManage
    */
   async getContractCode(address: Address): Promise<Uint8Array> {
     const account = await this.getAccount(address)
+    if (!account) {
+      return new Uint8Array(0)
+    }
     if (!account.isContract()) {
       return new Uint8Array(0)
     }
     const key = this._prefixCodeHashes
       ? concatBytes(CODEHASH_PREFIX, account.codeHash)
       : account.codeHash
-    // @ts-expect-error
-    const code = await this._trie._db.get(key)
-    return code ?? new Uint8Array(0)
-  }
 
-  /**
-   * Creates a storage trie from the primary storage trie
-   * for an account and saves this in the storage cache.
-   * @private
-   */
-  async _lookupStorageTrie(address: Address): Promise<Trie> {
-    // from state trie
-    const account = await this.getAccount(address)
-    const storageTrie = this._trie.copy(false)
-    storageTrie.root(account.storageRoot)
-    storageTrie.flushCheckpoints()
-    return storageTrie
+    const keyHex = bytesToHex(key)
+    if (keyHex in this._codeCache) {
+      return this._codeCache[keyHex]
+    } else {
+      // @ts-expect-error
+      const code = (await this._trie._db.get(key)) ?? new Uint8Array(0)
+      this._codeCache[keyHex] = code
+      return code
+    }
   }
 
   /**
    * Gets the storage trie for an account from the storage
    * cache or does a lookup.
    * @private
    */
-  async _getStorageTrie(address: Address): Promise<Trie> {
+  async _getStorageTrie(address: Address, account: Account): Promise<Trie> {
     // from storage cache
     const addressHex = bytesToHex(address.bytes)
-    let storageTrie = this._storageTries[addressHex]
-    if (storageTrie === undefined || storageTrie === null) {
-      // lookup from state
-      storageTrie = await this._lookupStorageTrie(address)
+    const storageTrie = this._storageTries[addressHex]
+    if (storageTrie === undefined) {
+      const storageTrie = this._trie.copy(false)
+      storageTrie.root(account.storageRoot)
+      storageTrie.flushCheckpoints()
```

Comment:
Hmm should it not push this `storageTrie` in `_storageTries` cache here?

Reply:
Oh, super good point, definitely, have added. 👍 

(will push combined later on)